sys/net/gnrc/application_layer/tftp/gnrc_tftp.c

```diff
@@ -0,0 +1,973 @@
+                                             GNRC_NETTYPE_UNDEF);
+
+    /* check if this is an client start */
+    if (!m) {
+        DEBUG("tftp: starting transaction as client\n");
+        return _tftp_send_start(ctxt, outbuf);
+    }
+    else if (m->type == TFTP_TIMEOUT_MSG) {
+        DEBUG("tftp: timeout occured\n");
+        if (++(ctxt->retries) > GNRC_TFTP_MAX_RETRIES) {
+            /* transfer failed due to lost peer */
+            DEBUG("tftp: peer lost\n");
+            gnrc_pktbuf_release(outbuf);
+            return FAILED;
+        }
+
+        /* increase the timeout for congestion control */
+        ctxt->block_timeout <<= 1;
+
+        /* the send message timed out, re-sending */
+        if (ctxt->dst_port == GNRC_TFTP_DEFAULT_DST_PORT) {
+            DEBUG("tftp: sending timed out, re-sending\n");
+            /* we are still negotiating resent, start */
+            return _tftp_send_start(ctxt, outbuf);
+        }
+        else {
+            DEBUG("tftp: last data or ack packet lost, resending\n");
+            /* we are sending / receiving data */
+            /* if we are reading resent the ACK, if writing the DATA */
+            return _tftp_send_dack(ctxt, outbuf, (ctxt->op == TO_RRQ) ? TO_ACK : TO_DATA);
+        }
+    }
+    else if (m->type != GNRC_NETAPI_MSG_TYPE_RCV) {
+        DEBUG("tftp: unknown message");
+        gnrc_pktbuf_release(outbuf);
+        return BUSY;
+    }
+
+    gnrc_pktsnip_t *pkt = (gnrc_pktsnip_t*)(m->content.ptr);
+
+    assert(pkt->next && pkt->next->type == GNRC_NETTYPE_UDP);
+    assert(pkt->next->next && pkt->next->next->type == GNRC_NETTYPE_IPV6);
+
+    uint8_t *data = (uint8_t*)pkt->data;
+    udp_hdr_t *udp = (udp_hdr_t*)pkt->next->data;
+    ipv6_hdr_t *ip = (ipv6_hdr_t*)pkt->next->next->data;
+
+    xtimer_remove(&(ctxt->timer));
+
+    switch (_tftp_parse_type(data)) {
+    case TO_RRQ:
+    case TO_WRQ: {
+        if (byteorder_ntohs(udp->dst_port) != GNRC_TFTP_DEFAULT_DST_PORT) {
+            /* not a valid start packet */
+            DEBUG("tftp: incoming packet on port %d dropped\n", byteorder_ntohs(udp->dst_port));
+            gnrc_pktbuf_release(outbuf);
+            return FAILED;
+        }
+
+        /* reinitialize the context with the current client */
+        _tftp_init_ctxt(&(ip->src), NULL, ctxt->start_cb, ctxt->data_cb,
+                        _tftp_parse_type(data), ctxt);
+
+        /* get the context of the client */
+        ctxt->dst_port = byteorder_ntohs(udp->src_port);
+        DEBUG("tftp: client's port is %"PRIu16"\n", ctxt->dst_port);
+        int offset = _tftp_decode_start(ctxt, data, outbuf);
+        DEBUG("tftp: offset after decode start = %i\n", offset);
+        if (offset < 0) {
+            DEBUG("tftp: there is no data?\n");
+            gnrc_pktbuf_release(outbuf);
+            return FAILED;
+        }
+
+        /* validate if the application accepts the filename and modes */
```

Comment:
The start callback doesn't get to know the transfer mode anymore, so this comment is not valid anymore. But I wonder if it would be a good idea to populate the requested mode to the start callback. What's deal with `netascii` and `octet`? But at least for `mail` this would make sense. On the other hand I'm not sure if someone would use `mail` anyway.